Investigate and handle alerts (#66)
appuio/openshift4-docs: docs/modules/ROOT/pages/how-tos/monitoring/handle_alerts.adoc

= Investigating and handling alerts

[abstract]
Looking at an alert can be overwhelming.
This is specially true when not familiar with how the alert ended up in your inbox.
This how to provides you some help on where to start.

> == Investigat an alert

Reviewer: ```suggestion
== Investigate an alert
```

. Read the message
+
Reading the message usually should give you a concise statement on what's wrong.
Most of the time, that message should be enough to grasp what action needs to be taken.
+
[NOTE]
====
There is no standard field defined for alert messages.
OpenShift 4 alerts build on top of https://github.com/prometheus-operator/kube-prometheus[kube-prometheus].
There `annotations.message` is used, but `annotations.summary` and `annotations.descriptions` are also used.
====

> . Lookout for labels and annotations

Reviewer: ```suggestion
. Look out for labels and annotations
```

+

> Alerts can have a set labels and annotations attached.

Reviewer: ```suggestion
Alerts can have a set of labels and annotations attached.
```

The name of the time series are of special interest here.
+
[NOTE]
====
The source link only works when clusters are publicly accessible.
Check out the cluster specific documentation if the link doesn't work.
There you should be able to find instructions on how to gain access.
====

. Understand the source of a time series
+
Sometimes looking at a time series, it's not obvious what it's all about.
In those cases, it helps to understand where it's coming from.
Each time series has a `job` label.
That label refers to the scrape job that brought that time series into the system.
Use that name to identify the source.
Have a look at the sources documentation to find details about the time series.

. Inspect targets
+
Should the job name not be enough to identify the source of a time series, you can have a look at the targets to find more details.
Navigate to _Status_ > _Target_ and search for the job name.
This will give you details of the scraped Kubernetes endpoint along with the associated service and pod.

== Handle an alert

An alert can be addressed in several ways:

=== Manual intervention

Sometimes, a manual intervention is the only thing required.

> This is fine as long the issue is expected to resurface again.

Reviewer: I think you meant this?
```suggestion
This is fine as long the issue is not expected to resurface again.
```

Fix the issue using your Kubernetes API client of choice (for example `kubectl`, `oc` or the OpenShift Console).

=== Change of configuration

The system is expected to show up again but can be prevented by tweaking the configuration.
Giving more resources, fixing a typo, changing a setting are all potential solutions.

[CAUTION]
====
Be cautious with throwing resources at a problem.
Sometimes reducing system load is the better way to go.
====

> Change the configuration within the Sync configuration hierarchy.

Reviewer 2: ```suggestion
Change the configuration within the Syn configuration hierarchy.
```

Do so on the Cluster, Tenant or global level as appropriate.

=== Change code

> Code can behave badly and triggers an alert.

Reviewer 2: ```suggestion
Code can behave badly and trigger an alert.
```

Identify the circumstances that led to the failure and file a bug report.
Pull requests with a fix are also welcome.

Other trigger for a change of code are missing features.

> A likely case for this would be a commodore component not allowing to configure certain aspects. 

Reviewer: ```suggestion
A likely case for this would be a Commodore component not allowing to configure certain aspects. 
```

File a feature request or better, submit a pull request.

=== Change of alert rule

If the system is working as expected then altering the alert rule might be the answer.
This most likely results in a change of code.

[NOTE]
====
It was discussed to allow altering alert rules via the configuration hierarchy.

> The decision was to not allow this for now.
> This would made it too easy.

Reviewer: ```suggestion
The decision was to not allow this for now since it would make this too easy.
```
